docs/designs/ps_design.md

```diff
@@ -1,375 +1,191 @@
 
-The interface of gradient queue could be like this:
-
-```python
-class GradientQueue(object):
-    def __init__(self):
-        self.grad_queue = queue.Queue()
-    
-    def get_gradient(self):
-        pass
-        
-    def put_gradient(self):
-        pass
-```
+The optimizer supports two kinds of parameter updating strategies: sync-SGD and async-SGD. In sync-SGD, the optimizer needs to wait for a certain number of gradients, and then apply the gradients to parameters. In async-SGD, the  `apply_gradient` function of optimizer inside will be called inside `push_gradient` RPC service directly.
 
 ### RPC Service
 
-PS provides RPC service for workers.
-
-Since each PS node will store a subset of the whole model. A worker will push/pull submodels to/from all PS nodes.
+PS provides RPC service for workers. There are three important events that PS should interact with workers:
 
-However, embedding table is initialized lazily in PS, worker should also send embedding table information to PS. We have to add another field to describe embedding table related information.
+* initialization of model parameters
+* pull model parameters
+* push gradients
 
-The model message is defined as following:
+**Initialization of model parameters**
 
-```proto
-
-message EmbeddingTableInfo{
-    string name = 1;
-    repeated int64 dim = 2;
-    string initializer = 3;
-}
-
-message Model {
-    int64 version = 1;
-    repeated Tensor variables = 2;
-    repeated EmbeddingTableInfo embedding_table_info = 3;
-}
-```
+After starting, PS does not contain any parameter. For model variables, ElasticDL should initialize them before training process. For embedding vectors, ElasticDL adopts lazy initialization, i.e. initialize them when they are needed in training process.
 
-So the RPC service will be defined as following:
+Since a single PS pod may not have enough memory for big model, workers are responsible for random initializing model variables. After initializing, workers push initialized model variables to corresponding PS pod. Here is the RPC call definition for pushing initialized model.
 
 ```proto
 service PServer{
     rpc push_model(Model) returns (google.protobuf.Empty);
-    rpc pull_variable(PullModelRequest) returns (PullModelResponse);
-    rpc pull_embedding_vector(Tensor) returns (Tensor);
-    rpc push_gradient(PushGradientRequest) returns (PushGradientResponse);
 }
 ```
 
-For the details of RPC service definition, please refer to next section.
+**Pull model parameters**
 
-The interfaces of PServer could be like this:
+Since ElasticDL saves parameters in PS, workers should pull parameters from PS in each iteration of training/evaluation process. 
 
+For model variables, a worker needs to pull model variables from all PS pods before the forward-pass.
```

Comment:
Does it mean for each iteration, each worker will make multiple PRC call to each PS code to get the non-embedding variables? I'm wondering it will cost much time to do it.

Reply:
The overall data transmitted does not change. Multiple RPC calls can be called concurrently from a thread pool. This will take less time.